code/modules/reagents/chemistry/equilibrium.dm

```diff
@@ -0,0 +1,272 @@
+			return FALSE //Not hot enough
+	else
+		if(holder.chem_temp > reaction.required_temp)
+			return FALSE //Not cold enough
+			
+	if(! ((holder.pH >= (reaction.optimal_pH_min - reaction.determin_pH_range)) && (holder.pH <= (reaction.optimal_pH_max + reaction.determin_pH_range)) ))//To prevent pointless reactions
+		return FALSE
+	return TRUE
+
+/* 
+* Check to make sure our input vars are sensible - is the holder overheated? does it have the required reagents? Does it have the required calalysts?
+*
+* If you're adding more checks for reactions, this is the proc to edit
+* otherwise, generally, don't call this directed except internally
+*/
+/datum/equilibrium/proc/check_reagent_properties()
+	//Have we exploded?
+	if(!holder.my_atom || holder.reagent_list.len == 0)
+		return FALSE
+	//Are we overheated?
+	if(holder.chem_temp > reaction.overheat_temp)
+		SSblackbox.record_feedback("tally", "chemical_reaction", 1, "[reaction.type] overheated reaction steps")
+		reaction.overheated(holder, src)
+
+	//set up catalyst checks
+	var/total_matching_catalysts = 0
+	var/total_matching_reagents = 0
+
+	//If the product/reactants are too impure
+	for(var/r in holder.reagent_list)
+		var/datum/reagent/R = r
+		if (R.purity < reaction.purity_min)//If purity is below the min, call the proc
+			SSblackbox.record_feedback("tally", "chemical_reaction", 1, "[reaction.type] overly impure reaction steps")
+			reaction.overly_impure(holder, src)
+		//this is done this way to reduce processing compared to holder.has_reagent(P)
+		for(var/P in reaction.required_catalysts)
+			var/datum/reagent/R0 = P
+			if(R0 == R.type)
+				total_matching_catalysts++
+
+		for(var/B in reaction.required_reagents)
+			var/datum/reagent/R0 = B
+			if(R0 == R.type) // required_reagents = list(/datum/reagent/consumable/sugar = 1) /datum/reagent/consumable/sugar
+				total_matching_reagents++
+	
+	if(!(total_matching_reagents == reaction.required_reagents.len))
+		return FALSE
+
+	if(!(total_matching_catalysts == reaction.required_catalysts.len))
+		return FALSE
+
+	//All good!
+	return TRUE
+
+/*
+* Calculates how much we're aiming to create
+*
+* Specifically calcuates multiplier, product_ratio, target_vol
+* Also checks to see if these numbers are sane, returns a TRUE/FALSE
+* Generally an internal proc
+*/
+/datum/equilibrium/proc/calculate_yield()
+	if(to_delete)
+		return FALSE
+	if(!reaction)
+		stack_trace("Tried to calculate an equlibrium for reaction [reaction.type], but there was no reaction set for the datum")
+		return FALSE
+	multiplier = INFINITY
+	product_ratio = 0
+	target_vol = 0
+	for(var/B in reaction.required_reagents)
+		multiplier = min(multiplier, round((holder.get_reagent_amount(B) / reaction.required_reagents[B]), CHEMICAL_QUANTISATION_LEVEL))
+	for(var/P in reaction.results)
+		target_vol += (reaction.results[P]*multiplier)
+		product_ratio += reaction.results[P]
+	if(target_vol == 0 || multiplier == INFINITY)
```

Comment:
It seems you are using `INFINITY` here for an invalid value. Use `null` instead, a non-numeric value.

Reply:
The INFINITY here is for the min() method, so I start at the max possible and itteratively go down, i.e.
```c
for(var/B in reaction.required_reagents)
		multiplier = min(multiplier, round((holder.get_reagent_amount(B) / reaction.required_reagents[B]), CHEMICAL_QUANTISATION_LEVEL))
```
The check is to sanity check, but thinking about I'm not sure it can actually get here with infinity. If it doesn't find a reagent, it should return 0.

Reply:
Alright, feel free to resolve.
👍